python/ray/tune/impl/tuner_internal.py

```diff
@@ -69,21 +73,59 @@ def __init__(
     ):
         from ray.train.trainer import BaseTrainer
 
-        # Restored from Tuner checkpoint.
+        # If no run config was passed to Tuner directly, use the one from the Trainer,
+        # if available
+        if not run_config and isinstance(trainable, BaseTrainer):
+            run_config = trainable.run_config
+
+        self._tune_config = tune_config or TuneConfig()
+        self._run_config = run_config or RunConfig()
+
+        # Restore from Tuner checkpoint.
         if restore_path:
-            trainable_ckpt = os.path.join(restore_path, _TRAINABLE_PKL)
-            with open(trainable_ckpt, "rb") as fp:
+            # Sync down from cloud storage if needed
+            synced, experiment_checkpoint_dir = self._maybe_sync_down_tuner_state(
+                restore_path
+            )
+            experiment_checkpoint_path = Path(experiment_checkpoint_dir)
+
+            if (
+                not (experiment_checkpoint_path / _TRAINABLE_PKL).exists()
+                or not (experiment_checkpoint_path / _TUNER_PKL).exists()
+            ):
+                raise RuntimeError(
+                    f"Could not find Tuner state in restore directory. Did you pass"
+                    f"the correct path (including experiment directory?) Got: "
+                    f"{restore_path}"
+                )
+
+            # Load trainable and tuner state
+            with open(experiment_checkpoint_path / _TRAINABLE_PKL, "rb") as fp:
                 trainable = pickle.load(fp)
 
-            tuner_ckpt = os.path.join(restore_path, _TUNER_PKL)
-            with open(tuner_ckpt, "rb") as fp:
+            with open(experiment_checkpoint_path / _TUNER_PKL, "rb") as fp:
                 tuner = pickle.load(fp)
                 self.__dict__.update(tuner.__dict__)
 
             self._is_restored = True
             self._trainable = trainable
-            self._experiment_checkpoint_dir = restore_path
             self._resume_config = resume_config
+
+            if not synced:
+                # If we didn't sync, use the restore_path local dir
+                self._experiment_checkpoint_dir = restore_path
+            else:
+                # If we synced, `experiment_checkpoint_dir` will contain a temporary
+                # directory. Create an experiment checkpoint dir instead and move
+                # our data there.
```

Comment:
Should this be done in `_maybe_sync_down_tuner_state` to avoid the extra step of using a temp dir?

Reply:
It's not straightforward - to create the final experiment checkpoint dir we need the trainable/tuner state (for the experiment name and checkpoint dir definition). Thus we need to sync --> restore state --> create dir. We can't move creating the dir to the sync step.

Reply:
I've moved the function into a separate method that makes it a bit more self-contained